src/csharp/Microsoft.Spark/ML/Feature/Base.cs

```diff
@@ -0,0 +1,334 @@
+
+        /// <summary>Checks whether a param is explicitly set or has a default value.</summary>
+        /// <param name="param">The <see cref="Param"/> to be checked.</param>
+        /// <returns>bool</returns>
+        public bool IsDefined(Param.Param param) => (bool)Reference.Invoke("isDefined", param);
+
+        /// <summary>
+        /// Tests whether this instance contains a param with a given name.
+        /// </summary>
+        /// <param name="paramName">The <see cref="Param"/> to be test.</param>
+        /// <returns>bool</returns>
+        public bool HasParam(string paramName) => (bool)Reference.Invoke("hasParam", paramName);
+
+        /// <summary>
+        /// Retrieves a <see cref="Param"/> so that it can be used to set the value of the
+        /// <see cref="Param"/> on the object.
+        /// </summary>
+        /// <param name="paramName">The name of the <see cref="Param"/> to get.</param>
+        /// <returns><see cref="Param"/> that can be used to set the actual value</returns>
+        public Param.Param GetParam(string paramName) =>
+            new Param.Param((JvmObjectReference)Reference.Invoke("getParam", paramName));
+
+        /// <summary>
+        /// Sets the value of a specific <see cref="Param"/>.
+        /// </summary>
+        /// <param name="param"><see cref="Param"/> to set the value of</param>
+        /// <param name="value">The value to use</param>
+        /// <returns>The object that contains the newly set <see cref="Param"/></returns>
+        public T Set<T>(Param.Param param, object value) =>
+            WrapAsType<T>((JvmObjectReference)Reference.Invoke("set", param, value));
+
+        /// <summary>
+        /// Clears any value that was previously set for this <see cref="Param"/>. The value is
+        /// reset to the default value.
+        /// </summary>
+        /// <param name="param">The <see cref="Param"/> to set back to its original value</param>
+        /// <returns>Object reference that was used to clear the <see cref="Param"/></returns>
+        public T Clear<T>(Param.Param param) =>
+            WrapAsType<T>((JvmObjectReference)Reference.Invoke("clear", param));
+
+        protected static T WrapAsType<T>(JvmObjectReference reference)
+        {
+            ConstructorInfo constructor = typeof(T)
+                .GetConstructors(BindingFlags.NonPublic | BindingFlags.Instance)
+                .Single(c =>
+                {
+                    ParameterInfo[] parameters = c.GetParameters();
+                    return (parameters.Length == 1) &&
+                        (parameters[0].ParameterType == typeof(JvmObjectReference));
+                });
+
+            return (T)constructor.Invoke(new object[] { reference });
+        }
+    }
+
+    /// <summary>
+    /// <see cref="ScalaPipelineStage"/> A stage in a pipeline, either an Estimator or a Transformer.
+    /// </summary>
+    public abstract class ScalaPipelineStage : Params
```

Comment:
Because those classes are calling reflection on the Scala side, and if one day there're plans to implement the real Transformer in .NET directly we can use `Transformer`. So a similar example would be JavaTransformer that's used in pyspark to keep track of reflection usage instead of direct Transformer.

Reply:
emmm yes we can also do that, though I don't know why pyspark calls it Javaxxx with the source code is in scala